line-auth/functions/index.js

```diff
@@ -0,0 +1,149 @@
+  const firebaseUid = `line:${lineMid}`;
+
+  // LINE's get user profile API endpoint
+  const getProfileOptions = generateLineApiRequest('https://api.line.me/v1/profile', lineAccessToken);
+
+  return admin.auth().getUser(firebaseUid).catch(error => {
+    // If user does not exist, fetch LINE profile and create a Firebase new user with it
+    if (error.code === 'auth/user-not-found') {
+      return rp(getProfileOptions).then(response => {
+        // Parse user profile from LINE's get user profile API response
+        const displayName = response.displayName;
+        const photoURL = response.pictureUrl;
+   
+        console.log('Create new Firebase user for LINE user mid = "', lineMid,'"');
+        // Create a new Firebase user with LINE profile and return it
+        return admin.auth().createUser({
+          uid: firebaseUid,
+          displayName: displayName,
+          photoURL: photoURL
+        });
+      });
+    }
+    // If error other than auth/user-not-found occurred, fail the whole login process
+    throw error;
+  });
+}
+
+/**
+ * Verify LINE access token and return a custom auth token allowing signing-in 
+ * the corresponding Firebase account.
+ *
+ * Here are the steps involved:
+ *  1. Verify with LINE server that a LINE access token is valid
+ *  2. Check if a Firebase user corresponding to the LINE user already existed.
+ *  If not, fetch user profile from LINE and generate a corresponding Firebase user.
+ *  3. Return a custom auth token allowing signing-in the Firebase account.
+ *
+ * @returns {Promise<string>} The Firebase custom auth token in a promise.
+ */
+function verifyLineToken(lineAccessToken) {
+  // Send request to LINE server for access token verification
+  const verifyTokenOptions = generateLineApiRequest('https://api.line.me/v1/oauth/verify', lineAccessToken);
+  var firebaseUid = '';
+
+  // STEP 1: Verify with LINE server that a LINE access token is valid
+  return rp(verifyTokenOptions)
+    .then(response => {
+      // Verify the token’s channelId match with my channelId to prevent spoof attack
+      // <IMPORTANT> As LINE's Get user profiles API response doesn't include channelID,
+      // you must not skip this step to make sure that the LINE access token is indeed
+      // issued for your channel.
+      //TODO: consider !== here
+      if (response.channelId != config.line.channelId)
+        return Promise.reject(new Error('LINE channel ID mismatched'));
+
+      // STEP 2: Access token validation succeeded, so look up the corresponding Firebase user
+      const lineMid = response.mid;
+      return getFirebaseUser(lineMid, lineAccessToken);
+    })
+    .then(userRecord => {
+      // STEP 3: Generate Firebase Custom Auth Token
+      const tokenPromise = admin.auth().createCustomToken(userRecord.uid);
+      tokenPromise.then(token => {
+        console.log('Created Custom token for UID "', userRecord.uid, '" Token:', token);
+      });
+      
+      return tokenPromise;
```

Comment:
I need my verifyLineToken function to return a promise which resolve with a Firebase custom auth token, so have to write it this way. I also can't have console.log inside exports.verifyToken because uid only exists in the context of verifyLineToken...